feat: Implement Group Formation Job (#68)

adinschmidt opened the pull request:
## Summary

This PR implements the Group Formation Job feature (issue #68), significantly **simplified based on PR review feedback**.

## Key Simplifications (from @jliu1016's review)

| Original Approach | Simplified Approach |
|-------------------|---------------------|
| Every-minute cron polling | Event-driven: schedule at `departureTime - 15 mins` |
| PostgreSQL advisory locks | Simple local tracking (MVP doesn't need distributed locking) |
| Complex multi-layer transaction isolation | Standard repository operations |
| Global trigger endpoint | Target-specific endpoints for itinerary/trip |
| Complex GTFS time string handling | Use existing entity relationships |

---

## Features

### Service: `GroupFormationService`
- `formGroupsForItinerary(id)` - Form groups for all trips in an itinerary
- `formGroupsForTrip(trip)` - Form groups for a specific trip

### Scheduler: `GroupFormationScheduler` 
- Event-driven scheduling at `departureTime - 15 minutes`
- Call `scheduleForItinerary(id)` when itinerary is created
- Uses NestJS `SchedulerRegistry` for dynamic job scheduling

### Endpoints
- `POST /group-formation/itinerary/:itineraryId` - Trigger for specific itinerary
- `POST /group-formation/trip/:tripId` - Trigger for specific trip
- `GET /group-formation/status` - Health check

---

## Business Rules

| Rule | Implementation |
|------|----------------|
| Group Size | 2-5 members (matches GO Transit group pass) |
| Itinerary Matching | Users with identical itineraries grouped together |
| Steward Requirement | Each group needs ≥1 user with `wantsToSteward=true` |
| Status Filter | Only `CHECKED_IN` bookings without a group |

---

## Algorithm

```
1. Get eligible bookings (CHECKED_IN, no group)
2. Group by itinerary hash (sorted trip IDs)
3. For each itinerary group:
   a. Check minimum size (≥2)
   b. Check steward availability
   c. Split into groups of 2-5
   d. Assign one steward per group
   e. Create TravelGroup, update bookings
```

---

## Files Changed

| File | Description |
|------|-------------|
| `group-formation.service.ts` | Core grouping logic (~350 lines → ~300 lines) |
| `group-formation.controller.ts` | REST endpoints with Swagger docs |
| `group-formation.scheduler.ts` | Event-driven scheduling (~200 lines → ~150 lines) |
| `group-formation.module.ts` | NestJS module |
| `group-formation.service.spec.ts` | 17 unit tests |
| `test-utils.ts` | Mock entity factories |
| `index.ts` | Barrel exports |

---

## Testing

```bash
npm run test -- --run src/group-formation/
# 17 tests passing
```

---

## Usage

```typescript
// When an itinerary is created:
await groupFormationScheduler.scheduleForItinerary(itinerary.id);

// Or manually trigger:
POST /group-formation/itinerary/abc-123
```

---

## Closes

- #68 (parent issue)
- #90, #91, #92, #93, #94, #95 (sub-issues)

adinschmidt committed:
feat: implement group formation job (#68)

Implements the Group Formation Job feature including all sub-issues:
- #90: Setup NestJS Schedule/Cron - installed @nestjs/schedule
- #91: Create GroupFormationService - core grouping business logic
- #92: Query trips departing soon - finds trips in 15-min window
- #93: Grouping algorithm - splits users into groups of 2-5
- #94: Persist TravelGroup and update bookings - database operations
- #95: Unit tests for Group Formation - 21 tests passing

Features:
- Cron job runs every minute to check for departing trips
- Manual trigger endpoint: POST /group-formation/trigger
- Status endpoint: GET /group-formation/status
- Groups users by identical itineraries (same set of trips)
- Requires at least one steward candidate per group
- Groups without stewards fail (MVP behavior)
- Group size constraints: [2, 5] members

Closes #68, #90, #91, #92, #93, #94, #95

adinschmidt committed:
feat: address high-priority PR review concerns

- Add PostgreSQL advisory locks for distributed locking
- Add per-trip transaction isolation with row-level locking
- Add configurable env vars (ENABLED, CRON_SCHEDULE, WINDOW_MINUTES, etc.)
- Use UTC consistently for time comparisons
- Add metrics (runDurationMs, stewardShortageIncidents)
- Improve status endpoint to show configuration
- Increase tests from 21 to 32

jliu1016 commented on backend/src/group-formation/group-formation.scheduler.ts:
```diff
@@ -0,0 +1,195 @@
+   *
+   * Default: Every minute (* * * * *)
+   */
+  @Cron(process.env.GROUP_FORMATION_CRON_SCHEDULE ?? '* * * * *')
```
Can we just schedule this to run based on information in the itineraries

jliu1016 commented on backend/src/group-formation/group-formation.scheduler.ts:
```diff
@@ -0,0 +1,195 @@
+ * - Can be disabled via GROUP_FORMATION_ENABLED=false
+ */
+@Injectable()
+export class GroupFormationScheduler implements OnModuleInit {
```
feel like we should just table this for after the demo.

jliu1016 commented on backend/src/group-formation/group-formation.controller.ts:
```diff
@@ -0,0 +1,114 @@
+import { AuthGuard } from '../modules/auth/auth.guard';
+
+/**
+ * Controller for manually triggering group formation.
```
this controller won't always be for manually triggering group function

jliu1016 commented on backend/src/group-formation/group-formation.controller.ts:
```diff
@@ -0,0 +1,114 @@
+    status: 500,
+    description: 'Internal server error during group formation',
+  })
+  async triggerGroupFormation(): Promise<GroupFormationRunResult> {
```
This should only trigger group formation on a specific itinerary

jliu1016 commented on backend/src/group-formation/group-formation.service.ts:
```diff
@@ -0,0 +1,627 @@
+   * Convert a Date to Metrolinx service ID format (YYYYMMDD).
+   * Uses local date since service IDs are typically based on local dates.
+   */
+  private toServiceIdFormat(date: Date): string {
```
use the existing util function gtfsDateStringToDate

jliu1016 commented on backend/src/group-formation/group-formation.service.ts:
```diff
@@ -0,0 +1,627 @@
+   * Convert a Date to GTFS time string format (HH:MM:SS) using UTC.
+   * This ensures consistent behavior regardless of server timezone.
+   */
+  private toGTFSTimeStringUTC(date: Date): string {
```
move to util function

jliu1016 commented on backend/src/group-formation/group-formation.service.ts:
```diff
@@ -0,0 +1,627 @@
+  ): Promise<TravelGroup | null> {
+    try {
+      // Get the next group number for this trip
+      const existingGroupCount = await txEm.count(TravelGroup, {
```
we shouldn't be interacting directly with travelgroups through EM - we should be injecting the repositories

jliu1016 commented on backend/src/group-formation/group-formation.service.ts:
```diff
@@ -0,0 +1,627 @@
+    );
+
+    // Get current time in GTFS format (HH:MM:SS) using UTC
+    const currentTimeStr = this.toGTFSTimeStringUTC(now);
```
a lot of this logic can be rolled into querying the repository, e.g. checking the departure window
const tripsInWindow = await this.tripRepository.find(
      {
        gtfsTrip: {
          serviceId: todayServiceId,
        },
        originStopTime: {
          departureTime: {
            $gte: currentTimeStr as GTFSTimeString,
            $lte: windowEndTimeStr as GTFSTimeString,
          },
        },
        bookings: {
          status: TripBookingStatus.CHECKED_IN,
          group: null,
        },
      },
      {
        populate: ['gtfsTrip', 'originStopTime', 'destinationStopTime'],
      },
    );

jliu1016 commented:
might want to look into using @Version() in mikroorm rather than implementing our own locks on stuff

adinschmidt committed:
refactor: simplify group formation based on PR feedback

Addresses comments from jliu1016's review:

1. Event-driven scheduling instead of cron polling
   - Schedule one-time jobs at (departureTime - 15 mins)
   - Triggered when itineraries are created
   - Removed every-minute cron job

2. Removed over-engineered distributed locking
   - No advisory locks (overkill for MVP demo)
   - Simple local tracking of scheduled trips

3. Use repositories consistently
   - Inject and use EntityRepository for all entities
   - Not using EntityManager directly

4. Controller accepts specific targets
   - POST /group-formation/itinerary/:itineraryId
   - POST /group-formation/trip/:tripId
   - Removed global trigger endpoint

5. Simplified service interface
   - formGroupsForItinerary(id)
   - formGroupsForTrip(trip)

Tests: 17 passing

adinschmidt committed:
Handle empty group formation with clear failure errors

adinschmidt commented:
Replying to earlier review comments from @jliu1016:

1) Scheduler should be event-driven, not cron: Updated to event-driven scheduling with one-time timeout at (departureTime – 15 min). For the demo this isn’t auto-wired to itinerary creation/check-in, so we’ll call `scheduleForItinerary`/`scheduleForTrip` manually (or use the manual trigger endpoints). Cron polling removed.

2) Over-engineering / table after demo: Simplified as suggested—no advisory/distributed locks, no extra isolation layers; using repository operations and local tracking only for MVP.

3) Controller should be itinerary-specific (not global trigger): Controller now only exposes `POST /group-formation/itinerary/:itineraryId` and `POST /group-formation/trip/:tripId`; global trigger was removed.

4) Trigger should be itinerary-specific: The itinerary endpoint now requires an itinerary ID and runs formation for just that itinerary’s trips.

5) Use existing GTFS date util: Removed the custom date/time formatting helpers in this flow; we rely on the trip’s service date and stop times directly.

6) Move GTFS time helper to util: Same—no custom GTFS time helper remains; we use the trip’s GTFS stop times directly.

7) Avoid raw EM for TravelGroup: Switched to injected repositories. `createTravelGroup` uses `travelGroupRepository.create` and `persistAndFlush`; other access is via repositories.

8) Query departure window via repository filters: The new approach schedules by trip departure time and filters eligibles via repository queries; no manual time-window scans.

9) Consider @Version instead of custom locks: Custom/advisory locks were removed for MVP; no locking implemented now. We can revisit @Version later if needed.

jliu1016 commented on backend/.env.example:
```diff
@@ -48,3 +48,25 @@ DB_NAME=postgres
+# Cron schedule for group formation (default: every minute)
+# Uses standard cron format: "seconds minutes hours day month weekday"
+# Example: "*/5 * * * *" for every 5 minutes
+GROUP_FORMATION_CRON_SCHEDULE=* * * * *
```
remove unused env vars

jliu1016 commented on backend/src/group-formation/group-formation.controller.ts:
```diff
@@ -0,0 +1,164 @@
+  })
+  @ApiResponse({ status: 401, description: 'Unauthorized' })
+  @ApiResponse({ status: 404, description: 'Itinerary not found' })
+  async formGroupsForItinerary(
```
Can we use the @Serialize() decorator (see other controllers)

adinschmidt replied:
done

jliu1016 commented on backend/src/group-formation/group-formation.controller.ts:
```diff
@@ -0,0 +1,164 @@
+  @ApiResponse({ status: 404, description: 'Itinerary not found' })
+  async formGroupsForItinerary(
+    @Param('itineraryId') itineraryId: string,
+  ): Promise<{ results: GroupFormationResult[] }> {
```
Please create zod dtos that get added to the shared packages

adinschmidt replied:
done

jliu1016 commented on backend/src/group-formation/group-formation.controller.ts:
```diff
@@ -0,0 +1,164 @@
+  ): Promise<GroupFormationResult> {
+    this.logger.log(`Group formation triggered for trip ${tripId}`);
+
+    const trip = await this.tripRepository.findOne({ id: tripId });
```
controllers should be lightweight and just call a servicee, they shouldn't use repositories.

jliu1016 commented on backend/src/group-formation/group-formation.controller.ts:
```diff
@@ -0,0 +1,164 @@
+      await this.groupFormationService.formGroupsForItinerary(itineraryId);
+
+    const totalGroupsFormed = results.reduce(
+      (sum, r) => sum + r.groupsFormed,
```
controllers should be lightweight and just call a servicee, they shouldn't use repositories.

adinschmidt replied:
removed `tripRepository` from this controller

jliu1016 commented on backend/src/group-formation/group-formation.service.ts:
```diff
@@ -0,0 +1,377 @@
+    this.logger.log(`Forming groups for itinerary ${itineraryId}`);
+
+    // Get the itinerary with its trip bookings
+    const itinerary = await this.itineraryRepository.findOne(
```
Logic here is wrong. What we want to do - is find all the users with the same itinerary (I slacked you the branch itineraryAggregation), from there we create travel groups of people

jliu1016 commented on backend/src/group-formation/group-formation.service.ts:
```diff
@@ -0,0 +1,377 @@
+    // Process each trip in the itinerary
+    for (const booking of itinerary.tripBookings.getItems()) {
+      const trip = booking.trip;
+      const result = await this.formGroupsForTrip(trip);
```
The current logic goes:
Finds itinerary user provided
Finds all trips
From all the trips, it tries to form group

In the formGroupsForTrip function it tries to find out which trips have the same itineraries.

It would be more performant to do this:
Finds itinerary user provided
Finds all itineraries that have the same hash (or we use a SQL view and group by, like I did in itineraryAggregation, in which case the db would find all the itineraries that have the same trips booked and we wouldn't need a hash to filter)
from those itineraries we would pull out all the users and trips needed to create the groups,

adinschmidt replied:
For the demo, isn't the current in-memory grouping preferred? It uses a simpler query and we don't expect the performance to be an issue

jliu1016 replied:
Even if this was done programmatically, it seems to make more sense to filter from itineraries level. I'm not sure if having a simpler sql query is enough justification either. This way of doing it just creates much more code to maintain than is necessary.

adinschmidt committed:
refactor: address PR #128 feedback

- Remove unused env vars from .env.example
- Move validation schemas to shared package (Zod)
- Refactor GroupFormationController to use @Serialize decorator
- Remove repository injection from controller, adding lookup method to service

adinschmidt committed:
refactor(group-formation): efficient itinerary grouping via SQL view

Commit:
add automatic trip hashing to itinerary

adinschmidt committed:
switch group formation to use itinerary tripHash for matching instead of the SQL view

jliu1016 commented on backend/src/group-formation/group-formation.service.ts:
```diff
@@ -0,0 +1,401 @@
+    // Process each trip in the itinerary
+    for (const booking of itinerary.tripBookings.getItems()) {
+      const trip = booking.trip;
+      const result = await this.formGroupsForTrip(trip, tripHash);
```
logic issue here, we want to create the exact same group on every trip booking

adinschmidt replied:
see new changes

adinschmidt committed:
chore: update main application entry and group formation module imports

- Added 'reflect-metadata' import to main.ts for enhanced metadata reflection.
- Included UsersModule in GroupFormationModule imports to facilitate user-related functionalities.
- Simplified OrmModule imports by removing unnecessary MikroOrmModule.forFeature call.

adinschmidt commented:
Functional flow completed:

- create user `gftest1765265832@example.com`
- Signed in to get a valid access token.
- Inserted the same user into the app DB (`users` table) so AuthGuard could resolve it.
- Called `POST /group-formation/itinerary/bbbbbbbb-bbbb-bbbb-bbbb-bbbbbbbbbbb1` with the token → **201** and groups formed:
  ```
  {"data":{"results":[{"tripId":"99999999-9999-9999-9999-999999999999","groupsFormed":1,"usersGrouped":2,"usersNotGrouped":0,"failedNoSteward":0}]}}
  ```
- DB verification:
  - `travel_group`: `004f6b06-88d1-4ed0-b204-78f77acd5c15`, status forming, steward `aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaa1`.
  - `trip_booking`: both seeded bookings now `grouped` and linked to that group.

adinschmidt committed:
refactor(group-formation): unify group membership across multi-leg journeys

Replace individual trip-based grouping with a coordinated approach:
1. Query itineraries by tripHash to discover route-matching travelers
2. Partition users via splitItinerariesIntoGroups respecting steward/size rules
3. Apply identical membership to each trip via createGroupsForTrip

Test validates that both legs of a two-trip itinerary share the same
grouped users with proper status transitions.

jliu1016 commented on backend/src/group-formation/group-formation.service.ts:
```diff
@@ -0,0 +1,593 @@
+      return [];
+    }
+
+    const tripHash = itinerary.tripHash ?? this.computeTripHash(itinerary);
```
should throw error if it's not there. should never be in this state.

jliu1016 commented on backend/src/group-formation/group-formation.service.ts:
```diff
@@ -0,0 +1,593 @@
+    const tripHash = itinerary.tripHash ?? this.computeTripHash(itinerary);
+    if (!tripHash) {
+      this.logger.warn(
+        `Itinerary ${itineraryId} missing trip hash; cannot group bookings`,
```
throw internal server error - should never happen systematically

jliu1016 commented on backend/src/group-formation/group-formation.service.ts:
```diff
@@ -0,0 +1,593 @@
+    const n = itineraries.length;
+
+    const minGroupsNeeded = Math.ceil(n / MAX_GROUP_SIZE);
+    const maxGroupsPossible = Math.floor(n / MIN_GROUP_SIZE);
```
let's make this a parameter with default option. We don't need to expose this parameter in the controller

jliu1016 commented on backend/src/group-formation/group-formation.service.ts:
```diff
@@ -0,0 +1,593 @@
+        populate: [
+          'tripBookings',
+          'tripBookings.trip',
+          'tripBookings.user',
```
mikroorm has some pretty good features that mean you don't need to do like 96

// finds all authors that have some book called `Foo`
const res1 = await em.find(Author, {
  books: { $some: { title: 'Foo' } },
});

// finds all authors that have no books called `Foo`
const res2 = await em.find(Author, {
  books: { $none: { title: 'Foo' } },
});

// finds all authors that have every book called `Foo`
const res3 = await em.find(Author, {
  books: { $every: { title: 'Foo' } },

jliu1016 commented on backend/src/group-formation/group-formation.controller.ts:
```diff
@@ -0,0 +1,167 @@
+      `Group formation complete for itinerary ${itineraryId}: ${totalGroupsFormed} groups formed`,
+    );
+
+    return { results };
```
for now we assume that the groups are the same for whole itinerary, only return one result and document this assumption.

(We always form groups for a whole trip, there and back. No need to return groups per trip. Frontend is agnostic to trip bookings and trips for now)

jliu1016 commented on backend/src/group-formation/group-formation.service.ts:
```diff
@@ -0,0 +1,593 @@
+    itineraries: Itinerary[],
+    stewardItineraries: Itinerary[],
+  ): Itinerary[][] {
+    const n = itineraries.length;
```
using the variable n here actually makes this code more unreadable

jliu1016 commented on backend/src/group-formation/group-formation.service.ts:
```diff
@@ -0,0 +1,593 @@
+    }
+
+    // Distribute extra stewards
+    for (let i = numGroups; i < stewardItineraries.length; i++) {
```
starts at "numgroups" because the next steward to be slotted would be the one who didn't need to become a steward.

jliu1016 commented on backend/src/group-formation/group-formation.service.ts:
```diff
@@ -0,0 +1,593 @@
+
+    const minGroupsNeeded = Math.ceil(n / MAX_GROUP_SIZE);
+    const maxGroupsPossible = Math.floor(n / MIN_GROUP_SIZE);
+    const numGroups = Math.min(maxGroupsPossible, stewardItineraries.length);
```
groupsBeingCreated? This name isn't very explanatory.
